# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/jit/MacroAssembler.cpp
# Commit: 5af60a416884
# Full Hash: 5af60a4168846b5be7576faecbb5050fa6d965cf
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2023-07-13 09:17:48
# Description:
#   Bug 1842617: Check for empty shape list in branchTestObjShapeList r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D183192
# ==============================================================================

diff -r 6bab1aefd0bd -r 5af60a416884 js/src/jit/MacroAssembler.cpp
--- a/js/src/jit/MacroAssembler.cpp	Wed Jul 12 16:24:01 2023 +0000
+++ b/js/src/jit/MacroAssembler.cpp	Wed Jul 12 16:39:14 2023 +0000
@@ -4449,6 +4449,7 @@
 
   Label done;
   Label* onMatch = cond == Assembler::Equal ? label : &done;
+  Label* onNoMatch = cond == Assembler::Equal ? &done : label;
 
   // Load the object's shape pointer into shapeScratch, and prepare to compare
   // it with the shapes in the list. The shapes are stored as private values so
@@ -4459,6 +4460,7 @@
   Address lengthAddr(shapeElements,
                      ObjectElements::offsetOfInitializedLength());
   load32(lengthAddr, endScratch);
+  branch32(Assembler::Equal, endScratch, Imm32(0), onNoMatch);
   BaseObjectElementIndex endPtrAddr(shapeElements, endScratch);
   computeEffectiveAddress(endPtrAddr, endScratch);
 
@@ -4483,8 +4485,8 @@
 
   if (cond == Assembler::NotEqual) {
     jump(label);
-    bind(&done);
-  }
+  }
+  bind(&done);
 }
 
 void MacroAssembler::branchTestObjCompartment(Condition cond, Register obj,
